### PR TITLE
Further improvements and additional performance tests for AzMath

### PR DESCRIPTION
**Pull request by kberg-amzn**
For convenience, lets certain operations that return outputs properly size the outputs to the correct dimensionalities rather than asserting.
Adds guards that protect against crashes if null (empty) vectors or matrices are operated on.
Distinguishes between L1 (Manhattan), and L2 (Euclidean) norms for VectorN norm calculations, and improves the performance of norm and dot product operations.
Adds a new set of benchmarks for AZ::VectorN and AZ::MatrixMxN

These benchmarks are timed on a 500 x 1,000 element matrix, a 1,000 x 500 element matrix, a 1,000 element vector and a 500 element vector.
`BM_MathMatrixMxN/VectorMatrixMultiply          107338 ns`
`BM_MathMatrixMxN/VectorMatrixMultiplyLeft      119668 ns`
`BM_MathMatrixMxN/MatrixMatrixMultiply        38846188 ns`
`BM_MathMatrixMxN/OuterProduct                  109747 ns`

These benchmarks are timed on vectors containing 4,000 elements.
`BM_MathVectorN/GetNormalized                     2002 ns`
`BM_MathVectorN/Dot                               1183 ns`
`BM_MathVectorN/NaiveDotBaseline                  2095 ns`
`BM_MathVectorN/GetMin                            2082 ns`
`BM_MathVectorN/GetMax                            2060 ns`
`BM_MathVectorN/GetClamp                          2445 ns`
`BM_MathVectorN/Sub                               1899 ns`
`BM_MathVectorN/Sum                               2022 ns`
`BM_MathVectorN/Mul                               2057 ns`
`BM_MathVectorN/Div                               1983 ns`
`BM_MathVectorN/GetAbs                            1601 ns`

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Math/VectorN.inl [hunk 1]
```diff
@@ -233,19 +233,19 @@ namespace AZ
...
 
-    AZ_MATH_INLINE float VectorN::GetLengthSq() const
+    AZ_MATH_INLINE float VectorN::L1Norm() const
     {
-        float lengthSquared = 0.0f;
+        AZ::Vector4 partialLengths = AZ::Vector4::CreateZero();
         for (AZStd::size_t i = 0; i < m_values.size(); ++i)
         {
-            lengthSquared += m_values[i].GetLengthSq();
+            partialLengths += m_values[i].GetAbs();
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Math/VectorN.inl:241]: Are the unused elements of the final Vector4 element is 0 when this function is run?

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Math/VectorN.h [hunk 2]
```diff
@@ -96,11 +96,11 @@ namespace AZ
...
 
-        //! Returns squared length of the vector.
-        float GetLengthSq() const;
+        //! Returns L1 norm (Manhattan distance) of the vector, full accuracy.
+        float L1Norm() const;
 
-        //! Returns length of the vector, full accuracy.
-        float GetLength() const;
+        //! Returns L2 norm (Euclidean distance) of the vector, full accuracy.
+        float L2Norm() const;
```
Comment (lsemp3d) [inline Code/Framework/AzCore/AzCore/Math/VectorN.h:103]: I like the change, just mildly concerned it doesn't follow the convention

**Review by lsemp3d — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Math/VectorN.inl (hunk 1 above)
Comment (kberg-amzn) [reply Code/Framework/AzCore/AzCore/Math/VectorN.inl:241]: Yes, we always ensure unused elements are zero.

File: Code/Framework/AzCore/AzCore/Math/VectorN.h (hunk 2 above)
Comment (kberg-amzn) [reply Code/Framework/AzCore/AzCore/Math/VectorN.h:103]: Yeah, the worlds of numerical analysis and games are colliding a bit here.. Distinguishing between different types of norms is really important in many industries, including machine learning, but quite unfamiliar to game devs who are only used to dealing with L2 norms..

**Review by burelc-amzn — APPROVED** [review]

**Pull request merged**